dali/pipeline/pipeline.cc

```diff
@@ -564,23 +587,28 @@ void Pipeline::SetupCPUInput(std::map<string, EdgeMeta>::iterator it,
   input_strs.first = "contiguous_" + input_strs.first;
 }
 
-void Pipeline::SetupGPUInput(std::map<string, EdgeMeta>::iterator it) {
+void Pipeline::SetupGPUInput(std::map<string, EdgeMeta>::iterator it, int logical_id) {
   if (it->second.has_gpu) return;
   OpSpec copy_to_dev_spec =
     OpSpec("MakeContiguous")
     .AddArg("device", "mixed")
     .AddInput(it->first, "cpu")
     .AddOutput(it->first, "gpu");
   // don't put it into op_specs_for_serialization_, only op_specs_
-  this->op_specs_.push_back(make_pair("__Copy_" + it->first, copy_to_dev_spec));
+  this->op_specs_.push_back({"__Copy_" + it->first, copy_to_dev_spec, logical_id});
   it->second.has_gpu = true;
 }
 
-void Pipeline::PrepareOpSpec(OpSpec *spec) {
+void Pipeline::PrepareOpSpec(OpSpec *spec, int logical_id) {
+  seed_for_logical_id_.resize(GetLogicalIdCount(), {0, false});
```

Comment:
I think the number of operators is small enough not to worry about performance of `PrepareOpSpec` or `Build`. A such, I'd suggest that you can use a map/unordered map and do away with the `was_set` thing and direct indexing.

Reply:
Done, due to how we do serialization.